File: 1510-stone-game-iv/1510-stone-game-iv.py
@cache
def dfs(x: int) -> bool:
    if not x: return False
    i = 1
    while i * i <= x:
        if not dfs(x - i * i):
            return True
        i += 1
    return False

class Solution:
    def winnerSquareGame(self, n: int) -> bool:  
        return dfs(n)

File: 1510-stone-game-iv/1510-stone-game-iv.java
class Solution {
    private int[] m;
    private boolean dfs(int x)
    {
        if(x == 0) return false;
        if(m[x] != 0) return m[x] == 1;
        for(int i = 1; i * i <= x; i++)
        {
            if(!dfs(x - i * i))
            {
                m[x] = 1;
                return true;
            }
        }

        m[x] = -1;
        return false;
    }

    public boolean winnerSquareGame(int n) {
        m = new int[n+1];
        return dfs(n);
    }
}